GuruFX/GuruFX.Core/Entities/Entity.cs

```diff
@@ -1,70 +1,235 @@
 ﻿using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
-using GuruFX.Core.Components;
 
 namespace GuruFX.Core.Entities
 {
 	public class Entity : IEntity
 	{
+		#region CONSTRUCTORS
+
+		public Entity()
+		{
+
+		}
+
+		public Entity(string name)
+		{
+			Name = name;
+		}
+
+		/// <summary>
+		/// Create a new Entity with list of Instantiated Components
+		/// </summary>
+		/// <param name="name"></param>
+		/// <param name="componentTypes"></param>
+		public Entity(string name, params Type[] componentTypes)
+			: this(name)
+		{
+			foreach (Type componentType in componentTypes)
+			{
+				CreateAndAddComponentOfType(componentType);
+			}
+		}
+
+		#endregion CONSTRUCTORS
+
+		#region PROPERTIES
+
 		/// <summary>
 		/// Dictionary of Components. Key = ComponentInstanceID, Value = Component
 		/// </summary>
-		Dictionary<Guid, IComponent> mComponents = new Dictionary<Guid, IComponent>();
+		public ConcurrentDictionary<Guid, IComponent> Components { get; set; } = new ConcurrentDictionary<Guid, IComponent>();
 
 		/// <summary>
 		/// Dictionary of Child Entities. Key = EntityInstanceID, Value = ChildEntity
 		/// </summary>
-		Dictionary<Guid, IEntity> mEntities = new Dictionary<Guid, IEntity>();
-
+		public ConcurrentDictionary<Guid, IEntity> Entities { get; set; } = new ConcurrentDictionary<Guid, IEntity>();
 
+		/// <summary>
+		/// Name of this Entity
+		/// </summary>
 		public string Name { get; set; }
 
+		/// <summary>
+		/// Instance GUID of this Entity
+		/// </summary>
 		public Guid InstanceID { get; set; } = Guid.NewGuid();
 
-		public Dictionary<Guid, IComponent> Components
+		/// <summary>
+		/// The Parent Entity. If Parent is NULL, then this entity is the root (or top-level) entity.
+		/// </summary>
+		public IEntity Parent { get; set; }
+
+		/// <summary>
+		/// Get the Root Entity (Root node of the Scene)
+		/// </summary>
+		public IEntity Root => Parent?.Parent ?? this;
```

Comment:
Shouldn't this be `this.Parent?.Root ?? this.Parent ?? this`? Otherwise isn't it just the grandparent?